### DISCUSSION TITLE
Custom Frontend

### DISCUSSION THREAD

**Discussion by unseen-ninja** [FirefoxCSS-Store/FirefoxCSS-Store.github.io#5] [Ideas]
Based on the mention of loading time and page size in this issue [Optimizations / Changes?](https://github.com/FirefoxCSS-Store/FirefoxCSS-Store.github.io/issues/2) I had a quick thought about building a simple and custom Frontend for this to further reduce the load.

From what I can gather the template that's currently in use actually ships a lot of bloated JS. I like the layout in general, but I think it could be improved by making the listings more readable (text on images can be tough) and also adding a "Download"-button directly to the listing as well.
Maybe a simple Card-like design could achieve this?

**Comment by Neikon** [comment]
a grid with this style would be nice 
![imagen](https://user-images.githubusercontent.com/1765319/105754552-82b43980-5f4a-11eb-983c-c9884e89ae90.png)

Maybe it is easy to do with microsoft resources, they call it "fabric core" https://developer.microsoft.com/en-us/fluentui#/get-started/web#fabric-core

**Comment by unseen-ninja** [comment]
That's similar to what I had in mind.
I threw together a little somthing: https://codier.io/creation/B13r4q3JO

This way the user would save one click to get to the desired result and it's easier to grasp the entirety of all themes with a quick glance.

**Reply by Neikon** [reply]
it looks nice.  tomorrow I'll try to do it, it's late here today

if we want to put a direct download button I think we would have to add these things as well:
* a section on how to install that particular theme
*package the theme in ".zip" format
*ask permission to redistribute the themes to the authors (I contacted all of them and most of them are happy with this initiative, the rest didn't answer XD. But right now we only link to they github we need to ask again to do that).

for now I think we should first change the web to this design with the "json" system but keeping only the link to the author's github. and when it's done we make the changes that the download button needs.

**Reply by unseen-ninja** [reply]
Oh, I've thought that we'd just link the button to the repo. We could maybe change the copy of the button; I've just named it this way because the button on the current website also is a download icon. Hosting the themes would result in a lot more work than is necessary.

But I like the idea of having a simple guide on how to install themes in general. We could explain the basic steps and whatever is special for themes should be specified in the repo anyway. If we really want to spice things up we could also have a subpage for each theme and parse the readme from the repo. :D

**Reply by Neikon** [reply]
Ah okay , I've thought you was talking real/direct download button. clarified then.

We could do a general guide/tutorial and put it or in the footer or like first theme or both places. in addition in each theme we can add the readme.md of their repo, if we can render well these text in markdown. that way we have a theme-specific tutorial that we don't have to write ourselves.

**Reply by unseen-ninja** [reply]
For now I'd say it's probably better to keep it simple. Just list the themes and link to the repos and have a general installation guide. The specific details are listed in the repo anyway. We can keep the README parsing in mind for later tho.

I've also tinkered around with the cards a little more. This is where I'm currently at: [https://codier.io/creation/ry8O_qaJu](https://codier.io/creation/ry8O_qaJu) c:
The description is only visible when you hover over the item. It feels less messy this way imo. What do y'all think?

**Reply by Neikon** [reply]
The last design you have passed on looks great. I was also trying to make some cards with fluent design but I couldn't do anything decent . Yours looks much better. 

I've added the generic tutorial to the first position.

**Reply by BeyondMagic** [reply]
@andreasgrafen just saw the last design now, it's perfect, it seems very complete, is it yet?

**Reply by unseen-ninja** [reply]
Yeah, it pretty much is. I'm glad that you guys like it. c:

I'm currently in the process of applying this to the actual site. Ofc there are a few other things that have to be done for the site, but expect a pull request by tomorrow. ^—^

**Reply by unseen-ninja** [reply]
Update: Unfortunately real life destoryed my plans for working on the frontend. But I'm sure that I'll have some time over the weekend. Sorry for the delay!

**Reply by Neikon** [reply]
Don't worry 😉

**Reply by unseen-ninja** [reply]
Alrighty! I've just pushed the bare bones version of the new layout. As always you can check out the [live demo](https://andreas.grafen.info/other/FFCSS/). c:

For now it's really just the simple card stack and a little bit of sugar on top. It can easily be extended with some more content on site. ^—^

**Reply by BeyondMagic** [reply]
It's great, just waiting for the code here ;)

**Reply by unseen-ninja** [reply]
What do you mean "waiting for the code"? Just check the [frontend branch](https://github.com/FirefoxCSS-Store/FirefoxCSS-Store.github.io/tree/frontend). D:

**Reply by BeyondMagic** [reply]
I forgot that this branch existed 😞. So you passed to npm, I see, you're thinking about switching the html file to pug parser? Since there's already gulp.

**Reply by unseen-ninja** [reply]
It happens. We can't alway have every little thing in mind. :D

That's actually just my default gulp setup I use all the time… I just didn't bother to remove the stuff we donn't need right now and since we have talked about pug in the past I figured that it wouldn't matter too much anyway.

**Reply by BeyondMagic** [reply]
I see, do you need any help in the branch now or it's better we work on it when you merge it to the main?

**Reply by unseen-ninja** [reply]
The basic functionality is done. I think we could merge it once everyone agrees with the changes.

Once that's done we could take a closer look at the pug ordeal? If we could get rid of the JS loading in favour of serving static files we could imprive loading times even more. c:
From what I was able to gather during my limited research it should be possible to compile everything with gulp and GH actions, no?

**Reply by Neikon** [reply]
I have detected something that may or may not be a bug. I have detected a similar behaviour in the last change of the web, in the demo you could see a similar error(html error related to faviconloader) but then everything worked fine. Maybe it's an unimportant thing that only happens in the demo. I don't know.
![imagen](https://user-images.githubusercontent.com/1765319/106401809-73d7f600-6426-11eb-94ef-df819c4c3f28.png)

in addition, could we make a dark version before implementing the new frontend? could we put a little button at the top right to switch from one version to another?

in other hand i find this https://github.com/marketplace/actions/execute-javascript-inline 

and this one https://github.com/peaceiris/actions-gh-pages 
right here https://github.com/peaceiris/actions-gh-pages#%EF%B8%8F-static-site-generators-with-nodejs  it says something that seems to me that could be useful. (But I tell you that this escapes me completely and I hardly know what you are talking about. I understand the purpose but not the tools)
I don't know if these are the github actions we need.

**Reply by unseen-ninja** [reply]
Oh, I might have missed to deploy the images folder… error should be fixed now. c:

A darkmode is easily doable and a theme switcher isn't a big thing to implement as well. I can take a look at this tomorrow. The big issue comes into play when we want to serve the users chosen theme when they re-visit the site because that would require us to store cookies; and… let's be honest, the whole GDPR and cookie law stuff is hella annoying.

The links you've posted look promising. That should help a lot. ^—^

**Reply by Neikon** [reply]
and if we try to read the user's configuration like duckduckgo.com for example. if you have your browser set to dark, the website loads the dark version by default. that would no longer need cockies, right?

**Reply by unseen-ninja** [reply]
We could use `prefers-color-scheme` to check for the users preferences **but** FF will default to _light_ when the user has set `privacy.resistFingerprinting` to `true`.

**Comment by Neikon** [comment]
this could be the dark version.
![imagen](https://user-images.githubusercontent.com/1765319/106403775-8bb47780-6430-11eb-9d8f-45dc40a55805.png)
```
:root {
 --body-background: #1c1c1c;
 --body-background-lighter: #333333;
 --body-background-darker: #161616;
 --font-colour: #eeeeee;
}
```
and
```
.card {
    box-shadow: 0 1px 3px rgba(0,0,0,0.5);
```

**Reply by unseen-ninja** [reply]
I've thrown together some CSS and JS to make the magic happen. Users can switch between the light and dark themes and it also respects the users prefered colour scheme when the site is first loaded. The [demo](https://andreas.grafen.info/other/FFCSS/#) is updated too. c:

It's almost 2:30 am over here… I feel like I've missed something important but I can't pinpoint what it might be.

**Reply by Neikon** [reply]
looks awesome

**Comment by alfaaarex** [comment]
What about an Electron App which can install themes automatically?

**Reply by unseen-ninja** [reply]
This is prob. worth it's own discussion thread.

In general this would require an installer that works on all major systems. This is a **huge** task! To ensure that this ages well it also most likely also needs quite some maintenance to ensure compatibility with OS updates etc. — I personally don't want to deal with stuff like that at all as the web can be complicated enough. :D

The theme data is publically available tho; so if someone is up for the task.. ¯\\\_(ツ)\_/¯
